### PR TITLE
Hashable type class

### PR DESCRIPTION
**Pull request by fehrenbach**
See #188.

### CODE REVIEW

**Comment by kritzcreek** [comment]
This looks like a great start! I'm assuming you didn't just come up with these hashing functions and took them from somewhere else? (The Haskell implementation? Some paper?)

I think it would be good to have a few comments referencing where these hashing functions are coming from, to have some form of rationale for them.

**Comment by fehrenbach** [comment]
The one for Number is used by immutable.js, among others.
Haskell uses identity for Int and fromEnum for "small" types.
The String and Array algorithm here is the same as Java, I think.

I can add comments. It'd be even better if someone did some actual research on the matter. We are in the fortunate position that Eq is homogeneous and there is no subtyping, so changing the algorithm is not a breaking change and can be done for any type individually.

**Comment by fehrenbach** [comment]
This now uses the indexed newtype as discussed in #188.

File: src/Data/Hashable.purs [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+-- | ```PureScript
+-- | (a == b) `implies` (hash a == hash b)
+-- | ```
+-- |
+-- | That is, unequal hash values are a safe approximation of
+-- | inequality. In other words, two objects whose hash values differ,
+-- | are never equal. The reverse is not necessarily true.
+-- |
+-- | Hash values produced by `hash` must not be relied upon to be
+-- | stable accross multiple executions of a program and should not be
```
Comment (LiamGoodacre) [inline src/Data/Hashable.purs]: Typo: `accross` -> `across`

Comment (fehrenbach) [reply src/Data/Hashable.purs]: Thanks!

**Comment by fehrenbach** [comment]
Thanks to @rightfold for pointing out the array hash problem.

**Comment by JordanMartinez** [comment]
@fehrenbach Should this PR be closed in favor of #198?

**Comment by fehrenbach** [comment]
Yeah, probably. I should get back to this at some point...

**Pull request closed without merging**